magento/magento2
# Add Inline Form Key Updater to Page Cache | Fix Issue #13746

### Description (*)

This fix ensures that form keys are updated on frontend pages as soon as the form_key input field is loaded in order to prevent delays resulting in the "Invalid Form Key. Please refresh the page" error. 

A script tag with a unique ID is being added and executed inline directly after each form_key input ensuring its value is updated as soon as possible.

An internally-developed extension using the same approach is running on one of our customer's production server preventing this issue from occurring. The issue was more frequently encountered on long category pages due to the fact that the HTML had to be downloaded in multiple packets.

**This fix is provided by [MageMontreal](https://magemontreal.com), a [Magento Business Solution Partner](https://partners.magento.com/portal/details/partner/id/2232).**

### Fixed Issue


- fixes magento/magento2#13746: Add to Cart Form wrong Form Key in FPC

### Manual testing scenarios (*)

1. Create a long category page with several visible products
2. Make sure FPC is enabled and visit the page once with a browser
3. Using another browser or incognito, visit the same page which is now cached
4. Try adding a product to cart without waiting for the page to load

### Comments

Before applying this fix, when a long page or a page requiring several resources was opened, the form keys on the page were not updated immediately as the below script hadn't been executed yet.

**Script:** Magento/PageCache/view/frontend/web/js/page-cache.js

As this fix makes sure that the form keys are each updated inline as soon as the content has been downloaded by the browser, the issue doesn't occur anymore.

**Note:** Uses pure JavaScript to ensure that the scripts are executed without having to wait for RequireJS or any dependencies.

### Functional Requirement
Form key input fields must be integrated to the form through the Magento\Framework\View\Element\FormKey block which is best practice, in order to benefit from this fix/improvement.

### Contribution checklist (*)
